src/binary_tree/sparse_binary_tree.rs

```diff
@@ -0,0 +1,515 @@
+use ::std::collections::HashMap;
+use ::std::fmt::Debug;
+use thiserror::Error;
+
+// TODO maybe use concurrency for tree construction if it is too slow on the larger trees
+
+/// Minimum tree height supported.
+pub static MIN_HEIGHT: u8 = 2;
+
+// ===========================================
+// Main structs and constructor.
+
+/// Fundamental structure of the tree, each element of the tree is a Node.
+/// The data contained in the node is completely generic, requiring only to have an associated merge function.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Node<C: Clone> {
+    coord: Coordinate,
+    content: C,
+}
+
+/// The generic content type must implement this trait to allow 2 sibling nodes to be combined to make a new parent node.
+pub trait Mergeable {
+    fn merge(left_sibling: &Self, right_sibling: &Self) -> Self;
+}
+
+/// Used to identify the location of a Node
+/// y is the vertical index (height) of the Node (0 being the bottom of the tree).
+/// x is the horizontal index of the Node (0 being the leftmost index).
+#[derive(PartialEq, Eq, Hash, Debug, Clone)]
+pub struct Coordinate {
+    y: u8, // from 0 to height
+    // TODO change to 2^256
+    x: u64, // from 0 to 2^y
+}
+
+/// Main data structure.
+/// All nodes are stored in a hash map, their index in the tree being the key.
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct SparseBinaryTree<C: Clone> {
+    root: Node<C>,
+    store: HashMap<Coordinate, Node<C>>,
+    height: u8,
+}
+
+/// A simpler version of the Node struct that is used by the calling code to pass leaves to the tree constructor.
+#[allow(dead_code)]
+pub struct InputLeafNode<C> {
+    pub content: C,
+    pub x_coord: u64,
+}
+
+// ===========================================
+// Constructors.
+
+impl<C: Mergeable + Default + Clone> SparseBinaryTree<C> {
+    /// Create a new tree given the leaves, height and the padding node creation function.
+    /// New padding nodes are given by a closure. Why a closure? Because creating a padding node may require context outside of this scope, where type C is defined, for example.
+    #[allow(dead_code)]
```

Comment:
Rust usually uses the prefix "build" for functions that construct an object and return a Result. So this function could just be called `build()`.

I also **highly** suggest splitting the algorithm up into smaller functions. The `for` loops and variable assignments are making me nervous.

I am assuming that this is what you wanted to parallelise?

You could even create a builder/factory type containing `height` and `nodes` fields, and then just pass in predefined values to replace default values or `None`.

I would suggest letting `InputLeafNode` derive the `Default` trait if you choose to do this.

This will probably do best in a separate file